corehq/apps/sso/tests/test_tasks.py

```diff
@@ -133,3 +141,116 @@ def tearDown(self):
+        self._create_key(expiration_date=target_date)
+        non_compliant_keys = get_keys_expiring_after(self.users, target_date)
+        self.assertSetEqual(set(non_compliant_keys), set())
+
+    def test_finds_inactive_keys(self):
+        key = self._create_key(expiration_date=None, is_active=False)
+        non_compliant_keys = get_keys_expiring_after(self.users, datetime.datetime(year=2024, month=3, day=10))
+        self.assertSetEqual(set(non_compliant_keys), {key})
+
+    def _create_key(self, expiration_date=None, is_active=True):
+        return HQApiKey.objects.create(
+            user=self.user, key='key', name='test-key', expiration_date=expiration_date, is_active=is_active)
+
+
+class EnforceKeyExpirationTaskTests(TestCase):
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.account = generator.get_billing_account_for_idp()
+
+    def test_updates_non_compliant_expiry_to_max(self):
+        idp = self._create_idp_for_domains(['test.com'], max_days_until_user_api_key_expiration=30)
+        user = self._create_user('test-user@test.com')
+        key = self._create_key_for_user(user, expiration_date=None)
+
+        current_time = datetime.datetime(year=2024, month=8, day=1)
+        with freeze_time(current_time):
+            num_updates = enforce_key_expiration_for_idp(idp)
+
+        updated_key = HQApiKey.objects.get(id=key.id)
+        self.assertEqual(updated_key.expiration_date, current_time + datetime.timedelta(days=30))
+        self.assertEqual(num_updates, 1)
+
+    def test_exits_when_no_maximum_date_exists(self):
+        idp = self._create_idp_for_domains(['test.com'], max_days_until_user_api_key_expiration=None)
+        user = self._create_user('test-user@test.com')
+        self._create_key_for_user(user, expiration_date=None)
+
+        current_time = datetime.datetime(year=2024, month=8, day=1)
+        with freeze_time(current_time):
+            num_updates = enforce_key_expiration_for_idp(idp)
+
+        self.assertEqual(num_updates, 0)
+
+    def test_integration_test(self):
+        idp = self._create_idp_for_domains(['test.com'], max_days_until_user_api_key_expiration=30)
+        user = self._create_user('test-user@test.com')
+        key = self._create_key_for_user(user, expiration_date=None)
+
+        current_time = datetime.datetime(year=2024, month=8, day=1)
+        with freeze_time(current_time):
+            update_sso_user_api_key_expiration_dates(idp.id)
+
+        updated_key = HQApiKey.objects.get(id=key.id)
+        self.assertEqual(updated_key.expiration_date, current_time + datetime.timedelta(days=30))
+
+    def _create_idp_for_domains(self, domains, max_days_until_user_api_key_expiration=30):
+        idp = generator.create_idp('test-idp', account=self.account)
+        idp.max_days_until_user_api_key_expiration = max_days_until_user_api_key_expiration
+        idp.save()
+        if not domains:
+            domains = []
```

Comment:
nit: unused code
```suggestion
```

Reply:
> I think its reasonable to provide defaults, even if those defaults are not currently used

nit: While I used to agree, my mind has changed over time about dynamic languages like Python, especially in large codebases like HQ. Providing a default when it's never used can make it hard to evaluate possible code paths that may have been evaluated when debugging errors. For me a higher value is less code is better than more.

Reply:
I removed this code in 0bf1ee44c812f2238ce0b4d8d6fa77e8b7fd2d76, but I think probably for slightly different reasons. I still have the unused default of `30` for `max_days_until_user_api_key_expiration`, because the intent of the function is to provide a helper function that lets the test highlight only the relevant properties of the new IDP. However, given that the name of the function included `domains` (`_create_idp_for_domains`), it didn't make sense to call the function without specifying the list of domains.